File: src/io/innofang/singleton/example/java/ThreadSafeSynchronizedSingleton.java
package io.innofang.singleton.example.java;

/**
 * Created by Inno Fang on 2017/8/12.
 */
public class ThreadSafeSynchronizedSingleton {

    private static ThreadSafeSynchronizedSingleton instance;

    private ThreadSafeSynchronizedSingleton() {}

    public static synchronized ThreadSafeSynchronizedSingleton getInstance() {
        if (null == instance) {
            instance = new ThreadSafeSynchronizedSingleton();
        }
        return instance;
    }

}

File: src/io/innofang/singleton/example/kotlin/ThreadSafeSynchronizedSingleton.kt
package io.innofang.singleton.example.kotlin

/**
 * Created by Inno Fang on 2017/8/12.
 */
class ThreadSafeSynchronizedSingleton {

    companion object {
        private var instance: ThreadSafeSynchronizedSingleton? = null

        @Synchronized
        fun get(): ThreadSafeSynchronizedSingleton {
            if (null == instance) {
                instance = ThreadSafeSynchronizedSingleton()
            }
            return instance!!
        }
    }
}